File: construction_v1/report/ai_automation_analysis/ai_automation_analysis.py
"""
AI Automation Analysis - Native ERPNext Script Report

Comprehensive analysis of AI automation events, after-hours ticket handling,
document validation, data quality, and system health using native ERPNext doctypes
where applicable and custom doctypes for AI-specific metrics.

Leverages native Frappe/ERPNext doctypes:
- Scheduled Job Log: Automation execution tracking
- Error Log: System error monitoring
- Activity Log: User activity and audit trail
- Communication: Email/notification tracking

Includes WorkCom AI integration for intelligent insights.
"""

import json
from datetime import date, time, timedelta
from typing import Any, Dict, List, Optional, Tuple

import frappe
from frappe import _
from frappe.utils import getdate, get_first_day, get_last_day, now_datetime, get_datetime, cint, flt
from construction_v1.report.report_utils import get_period_dates

# Import business hours utilities
from construction_v1.business_utils import is_business_hours, get_business_hours


def execute(filters: Optional[Dict[str, Any]] = None) -> Tuple:
    """Main entry point for Script Report.

    Returns:
        Tuple of (columns, data, message, chart, report_summary, skip_total_row)
    """
    filters = frappe._dict(filters or {})
    get_period_dates(filters)

    columns = get_columns()
    data, summary_data = get_data(filters)
    chart = get_chart_data(summary_data)
    report_summary = get_report_summary(summary_data)

    return columns, data, None, chart, report_summary, False




def get_columns() -> List[Dict[str, Any]]:
    """Define report columns for scheduled job log data (native ERPNext)."""
    return [
        {"fieldname": "name", "label": "Log ID", "fieldtype": "Link", "options": "Scheduled Job Log", "width": 180},
        {"fieldname": "creation", "label": "Timestamp", "fieldtype": "Datetime", "width": 160},
        {"fieldname": "status", "label": "Status", "fieldtype": "Data", "width": 100},
        {"fieldname": "scheduled_job_type", "label": "Job Type", "fieldtype": "Link", "options": "Scheduled Job Type", "width": 200},
        {"fieldname": "is_after_hours", "label": "After Hours", "fieldtype": "Check", "width": 90},
        {"fieldname": "details", "label": "Details", "fieldtype": "Data", "width": 250},
    ]


def _is_business_hours(ts) -> bool:
    """Check if timestamp is within business hours using centralized utility."""
    return is_business_hours(ts)


def get_data(filters: frappe._dict) -> Tuple[List[Dict[str, Any]], Dict[str, Any]]:
    """Fetch scheduled job logs (native ERPNext) and compute summary metrics."""
    df = getdate(filters.date_from)
    dt = getdate(filters.date_to)

    # Build conditions for Scheduled Job Log (native Frappe doctype)
    conditions = ["sjl.creation BETWEEN %(df)s AND %(dt)s"]
    values = {"df": df, "dt": dt}

    if filters.get("status"):
        conditions.append("sjl.status = %(status)s")
        values["status"] = filters.status

    if filters.get("job_type"):
        conditions.append("sjl.scheduled_job_type LIKE %(job_type)s")
        values["job_type"] = f"%{filters.job_type}%"

    where_clause = " AND ".join(conditions)

    # Fetch scheduled job logs (native ERPNext doctype)
    logs = frappe.db.sql(f"""
        SELECT
            sjl.name, sjl.creation, sjl.status,
            sjl.scheduled_job_type, sjl.details
        FROM `tabScheduled Job Log` sjl
        WHERE {where_clause}
        ORDER BY sjl.creation DESC
        LIMIT 5000
    """, values, as_dict=True)

    # Build data rows and compute summary
    data = []
    summary = {
        "total_automations": 0,
        "success": 0,
        "failed": 0,
        "after_hours_count": 0,
        "by_job_type": {},
    }

    for log in logs:
        summary["total_automations"] += 1
        status = (log.get("status") or "").lower()

        if status == "complete":
            summary["success"] += 1
        elif status == "failed":
            summary["failed"] += 1

        is_after = not _is_business_hours(log.get("creation"))
        if is_after:
            summary["after_hours_count"] += 1

        # Track by job type
        job_type = log.get("scheduled_job_type") or "Unknown"
        summary["by_job_type"][job_type] = summary["by_job_type"].get(job_type, 0) + 1

        data.append({
            "name": log.get("name"),
            "creation": log.get("creation"),
            "status": log.get("status"),
            "scheduled_job_type": job_type,
            "is_after_hours": 1 if is_after else 0,
            "details": (log.get("details") or "")[:100],
        })

    # Compute additional summary metrics
    summary["success_rate"] = (
        round(summary["success"] / summary["total_automations"] * 100, 1)
        if summary["total_automations"] else 0
    )

    # Get after-hours ticket metrics from Unified Inbox Conversation
    after_hours = _get_after_hours_metrics(df, dt)
    summary.update(after_hours)

    # Get document validation metrics
    doc_validation = _get_document_validation_metrics(df, dt)
    summary.update(doc_validation)

    # Get data quality and AI failure metrics
    data_quality = _get_data_quality_metrics(df, dt)
    summary.update(data_quality)

    # Get system health metrics
    system_health = _get_system_health()
    summary.update(system_health)

    return data, summary


def _get_after_hours_metrics(df: date, dt: date) -> Dict[str, Any]:
    """Get after-hours ticket metrics from Unified Inbox Conversation."""
    convs = frappe.get_all(
        "Unified Inbox Conversation",
        filters={"creation_time": ["between", [df, dt]]},
        fields=["creation_time", "ai_handled", "platform"],
        limit=100000,
    )

    after_hours_convs = [c for c in convs if not _is_business_hours(c.creation_time)]
    total_after = len(after_hours_convs)
    ai_handled = sum(1 for c in after_hours_convs if c.ai_handled)

    by_platform: Dict[str, int] = {}
    for c in after_hours_convs:
        p = c.platform or "Unknown"
        by_platform[p] = by_platform.get(p, 0) + 1

    return {
        "after_hours_tickets": total_after,
        "after_hours_ai_handled": ai_handled,
        "after_hours_ai_rate": round(ai_handled / total_after * 100, 1) if total_after else 0,
        "after_hours_by_platform": by_platform,
    }


def _get_document_validation_metrics(df: date, dt: date) -> Dict[str, Any]:
    """Get document validation metrics."""
    docs = frappe.get_all(
        "Document Validation",
        filters={"validation_date": ["between", [df, dt]]},
        fields=["validation_status"],
        limit=100000,
    )

    total = len(docs)
    status_counts: Dict[str, int] = {}
    for d in docs:
        s = (d.validation_status or "Unknown").title()
        status_counts[s] = status_counts.get(s, 0) + 1

    failed = status_counts.get("Failed", 0)
    warnings = status_counts.get("Passed With Warnings", 0)

    return {
        "doc_validations_total": total,
        "doc_validations_failed": failed,
        "doc_validations_warnings": warnings,
        "doc_validation_status_counts": status_counts,
    }


def _get_data_quality_metrics(df: date, dt: date) -> Dict[str, Any]:
    """Get data quality issues and AI failure metrics."""
    try:
        from construction_v1.construction_v1.api.integration_monitoring import get_data_quality_metrics
        snapshot = get_data_quality_metrics() or {}
    except Exception:
        snapshot = {}

    issues = snapshot.get("quality_issues") or []
    issues_total = sum(i.get("count", 0) for i in issues)

    # Get AI failures from Failed Query Log
    failures = frappe.get_all(
        "Failed Query Log",
        filters={"creation": ["between", [df, dt]]},
        fields=["failure_reason"],
        limit=100000,
    )
    ai_failures_total = len(failures)
    by_reason: Dict[str, int] = {}
    for f in failures:
        r = f.failure_reason or "Unknown"
        by_reason[r] = by_reason.get(r, 0) + 1

    return {
        "data_quality_issues_total": issues_total,
        "data_quality_issues": issues,
        "ai_failures_total": ai_failures_total,
        "ai_failures_by_reason": by_reason,
    }


def _get_system_health() -> Dict[str, Any]:
    """Get system health metrics."""
    try:
        from construction_v1.construction_v1.production.monitoring_analytics_system import get_monitoring_system
        health = get_monitoring_system().calculate_system_health()
    except Exception:
        frappe.log_error(frappe.get_traceback(), "AI Automation Analysis system health error")
        health = {"overall_score": 0, "status": "unknown", "factors": {}}

    score = float(health.get("overall_score") or 0)
    status = health.get("status") or "unknown"

    return {
        "system_health_score": score,
        "system_health_status": status,
        "system_health_factors": health.get("factors") or {},
    }


def get_chart_data(summary: Dict[str, Any]) -> Dict[str, Any]:
    """Build primary automation status chart."""
    return {
        "data": {
            "labels": ["Success", "Failed"],
            "datasets": [{
                "name": "Automations",
                "values": [
                    summary.get("success", 0),
                    summary.get("failed", 0),
                ]
            }]
        },
        "type": "bar",
        "colors": ["#28a745", "#dc3545"],
    }


def get_report_summary(summary: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Build report summary cards."""
    health_score = summary.get("system_health_score", 0)
    health_indicator = "green" if health_score >= 80 else "orange" if health_score >= 60 else "red"

    return [
        {"value": summary.get("total_automations", 0), "label": "Automations", "datatype": "Int"},
        {"value": f"{summary.get('success_rate', 0):.1f}%", "label": "Success Rate", "datatype": "Data", "indicator": "green"},
        {"value": summary.get("after_hours_tickets", 0), "label": "After-Hours Tickets", "datatype": "Int", "indicator": "blue"},
        {"value": f"{summary.get('after_hours_ai_rate', 0):.1f}%", "label": "After-Hours AI Rate", "datatype": "Data"},
        {"value": summary.get("doc_validations_failed", 0), "label": "Doc Validation Fails", "datatype": "Int", "indicator": "orange"},
        {"value": summary.get("data_quality_issues_total", 0), "label": "Data Quality Issues", "datatype": "Int"},
        {"value": summary.get("ai_failures_total", 0), "label": "AI Failures", "datatype": "Int", "indicator": "red"},
        {"value": f"{health_score:.1f}%", "label": "System Health", "datatype": "Data", "indicator": health_indicator},
    ]


@frappe.whitelist()
def get_ai_insights(filters: str, query: str) -> Dict[str, Any]:
    """Return WorkCom-style insights for the AI Automation Analysis report."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})
    _ensure_dates(filters)

    df = getdate(filters.date_from)
    dt = getdate(filters.date_to)

    # Get current data
    _, summary = get_data(filters)

    # Get historical reports for trend analysis
    history = []
    try:
        history = frappe.get_all(
            "AI Automation Report",
            filters={"period_type": filters.get("period_type", "Monthly")},
            fields=[
                "name", "date_from", "date_to", "total_ai_automations",
                "after_hours_tickets", "after_hours_ai_handled",
                "invalid_documents_total", "invalid_documents_failed",
                "data_quality_issues_total", "ai_failures_total",
                "system_health_score", "system_health_status"
            ],
            order_by="date_from desc",
            limit=12,
        )
    except Exception:
        pass

    context = {
        "window": {
            "period_type": filters.get("period_type", "Monthly"),
            "from": str(df),
            "to": str(dt),
        },
        "current": {
            "total_ai_automations": summary.get("total_automations", 0),
            "success_count": summary.get("success", 0),
            "failed_count": summary.get("failed", 0),
            "success_rate": summary.get("success_rate", 0),
            "after_hours_tickets": summary.get("after_hours_tickets", 0),
            "after_hours_ai_handled": summary.get("after_hours_ai_handled", 0),
            "after_hours_ai_rate": summary.get("after_hours_ai_rate", 0),
            "doc_validations_total": summary.get("doc_validations_total", 0),
            "doc_validations_failed": summary.get("doc_validations_failed", 0),
            "data_quality_issues_total": summary.get("data_quality_issues_total", 0),
            "ai_failures_total": summary.get("ai_failures_total", 0),
            "system_health_score": summary.get("system_health_score", 0),
            "system_health_status": summary.get("system_health_status", ""),
        },
        "history": history,
    }

    try:
        from construction_v1.services.enhanced_ai_service import EnhancedAIService
        ai = EnhancedAIService()
        answer = ai.generate_ai_automation_report_insights(query=query, context=context)
        return {"insights": answer}
    except Exception:
        frappe.log_error(frappe.get_traceback(), "AI Automation Analysis AI Insights Error")
        return {
            "insights": (
                "AI insights are temporarily unavailable. Please ask your system "
                "administrator to configure WorkCom/OpenAI settings in Enhanced AI Settings."
            )
        }


@frappe.whitelist()
def get_automation_status_chart(filters: str) -> Dict[str, Any]:
    """Get automation success/failure chart data."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})
    _ensure_dates(filters)
    _, summary = get_data(filters)

    return {
        "data": {
            "labels": ["Success", "Failed"],
            "datasets": [{"name": "Count", "values": [
                summary.get("success", 0),
                summary.get("failed", 0),
            ]}]
        },
        "type": "bar",
        "colors": ["#28a745", "#dc3545"],
    }


@frappe.whitelist()
def get_after_hours_chart(filters: str) -> Dict[str, Any]:
    """Get after-hours tickets by platform chart data."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})
    _ensure_dates(filters)
    _, summary = get_data(filters)

    platform_counts = summary.get("after_hours_by_platform", {})
    plats = list(platform_counts.keys())
    counts = [platform_counts[p] for p in plats]

    return {
        "data": {
            "labels": plats if plats else ["No Data"],
            "datasets": [{"name": "After-Hours Tickets", "values": counts if counts else [0]}]
        },
        "type": "bar",
        "colors": ["#5e64ff"],
    }


@frappe.whitelist()
def get_document_validation_chart(filters: str) -> Dict[str, Any]:
    """Get document validation status chart data."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})
    _ensure_dates(filters)
    _, summary = get_data(filters)

    status_counts = summary.get("doc_validation_status_counts", {})
    labels = list(status_counts.keys())
    values = list(status_counts.values())

    return {
        "data": {
            "labels": labels if labels else ["No Data"],
            "datasets": [{"name": "Validations", "values": values if values else [0]}]
        },
        "type": "pie",
        "colors": ["#28a745", "#ffa00a", "#dc3545", "#5e64ff"],
    }


@frappe.whitelist()
def get_data_quality_chart(filters: str) -> Dict[str, Any]:
    """Get data quality issues chart data."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})
    _ensure_dates(filters)
    _, summary = get_data(filters)

    issues = summary.get("data_quality_issues", [])
    labels = [i.get("type", "Unknown") for i in issues]
    values = [i.get("count", 0) for i in issues]

    return {
        "data": {
            "labels": labels if labels else ["No Data"],
            "datasets": [{"name": "Data Issues", "values": values if values else [0]}]
        },
        "type": "bar",
        "colors": ["#ffa00a"],
    }


@frappe.whitelist()
def get_ai_failures_chart(filters: str) -> Dict[str, Any]:
    """Get AI failures by reason chart data."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})
    _ensure_dates(filters)
    _, summary = get_data(filters)

    by_reason = summary.get("ai_failures_by_reason", {})
    labels = list(by_reason.keys())
    values = list(by_reason.values())

    return {
        "data": {
            "labels": labels if labels else ["No Data"],
            "datasets": [{"name": "AI Failures", "values": values if values else [0]}]
        },
        "type": "bar",
        "colors": ["#dc3545"],
    }


@frappe.whitelist()
def get_system_health_chart(filters: str) -> Dict[str, Any]:
    """Get system health gauge chart data."""
    filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else filters or {})
    _ensure_dates(filters)
    _, summary = get_data(filters)

    score = summary.get("system_health_score", 0)

    return {
        "data": {
            "labels": ["Health", "Remaining"],
            "datasets": [{"name": "System Health", "values": [round(score, 1), max(0, 100 - round(score, 1))]}]
        },
        "type": "pie",
        "colors": ["#28a745", "#e9ecef"],
    }


@frappe.whitelist()
def get_automation_trend_chart(filters: str = None, months: int = 6) -> Dict[str, Any]:
    """Get automation trend chart over months using native Scheduled Job Log."""
    from frappe.utils import add_months

    labels = []
    success_values = []
    failed_values = []
    anchor = getdate()

    for i in range(months - 1, -1, -1):
        month_start = get_first_day(add_months(anchor, -i))
        if i > 0:
            month_end = get_first_day(add_months(anchor, -i + 1)) - timedelta(days=1)
        else:
            month_end = anchor

        # Count scheduled job logs in this month (native Frappe doctype)
        logs = frappe.get_all(
            "Scheduled Job Log",
            filters={"creation": ["between", [month_start, month_end]]},
            fields=["status"],
            limit=100000,
        )

        success = sum(1 for l in logs if (l.status or "").lower() == "complete")
        failed = sum(1 for l in logs if (l.status or "").lower() == "failed")

        labels.append(month_start.strftime("%b %Y"))
        success_values.append(success)
        failed_values.append(failed)

    return {
        "data": {
            "labels": labels,
            "datasets": [
                {"name": "Complete", "values": success_values},
                {"name": "Failed", "values": failed_values}
            ]
        },
        "type": "line",
        "colors": ["#28a745", "#dc3545"],
    }

File: construction_v1/report/ai_automation_analysis/ai_automation_analysis.js
/**
 * AI Automation Analysis - Native ERPNext Script Report Frontend
 *
 * Comprehensive AI automation analysis with filters, charts, and WorkCom AI integration.
 * Uses native Scheduled Job Log as the primary data source with aggregated metrics.
 */

frappe.query_reports["AI Automation Analysis"] = {
    filters: [
        {
            fieldname: "period_type",
            label: __("Period Type"),
            fieldtype: "Select",
            options: construction_v1.report_utils.period_type_options,
            default: "Monthly",
            reqd: 1,
            on_change: function () {
                construction_v1.report_utils.handle_period_change();
            }
        },
        {
            fieldname: "date_from",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.add_days(frappe.datetime.month_start(), -1),
            reqd: 1
        },
        {
            fieldname: "date_to",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1
        },
        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: "\nScheduled\nComplete\nFailed"
        },
        {
            fieldname: "job_type",
            label: __("Job Type"),
            fieldtype: "Link",
            options: "Scheduled Job Type"
        }
    ],

    onload: function (report) {
        // Add WorkCom AI button
        report.page.add_inner_button(__("Ask WorkCom"), function () {
            show_WorkCom_dialog(report);
        }, __("AI Insights"));

        // Add chart buttons
        report.page.add_inner_button(__("Automation Status"), function () {
            show_automation_status_chart(report);
        }, __("Charts"));

        report.page.add_inner_button(__("After-Hours Tickets"), function () {
            show_after_hours_chart(report);
        }, __("Charts"));

        report.page.add_inner_button(__("Document Validation"), function () {
            show_document_validation_chart(report);
        }, __("Charts"));

        report.page.add_inner_button(__("Data Quality Issues"), function () {
            show_data_quality_chart(report);
        }, __("Charts"));

        report.page.add_inner_button(__("AI Failures"), function () {
            show_ai_failures_chart(report);
        }, __("Charts"));

        report.page.add_inner_button(__("System Health"), function () {
            show_system_health_chart(report);
        }, __("Charts"));

        report.page.add_inner_button(__("Automation Trend"), function () {
            show_trend_chart(report);
        }, __("Charts"));
    },

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "execution_status" && data) {
            let status = (data.execution_status || "").toLowerCase();
            let cls = "gray";
            if (status === "success") cls = "green";
            else if (status === "failed" || status === "error") cls = "red";
            value = `<span class="indicator-pill ${cls}">${value}</span>`;
        }

        if (column.fieldname === "is_after_hours" && data) {
            if (data.is_after_hours) {
                value = `<span class="indicator-pill orange">Yes</span>`;
            } else {
                value = `<span class="indicator-pill gray">No</span>`;
            }
        }

        return value;
    }
};

function show_WorkCom_dialog(report) {
    let d = new frappe.ui.Dialog({
        title: __("Ask WorkCom - AI Automation Analytics Assistant"),
        fields: [
            {
                fieldname: "query",
                label: __("Your Question"),
                fieldtype: "Small Text",
                reqd: 1,
                placeholder: __("e.g., What are the key insights from AI automation this period?")
            },
            {
                fieldname: "response_section",
                fieldtype: "Section Break",
                label: __("WorkCom's Response")
            },
            {
                fieldname: "response",
                fieldtype: "HTML",
                options: '<div class="WorkCom-response" style="min-height:120px;padding:12px;background:#f5f7fa;border-radius:4px;"><em>Ask a question to get AI-powered automation insights...</em></div>'
            }
        ],
        primary_action_label: __("Ask WorkCom"),
        primary_action: function (values) {
            let $response = d.$wrapper.find(".WorkCom-response");
            $response.html('<div class="text-muted"><i class="fa fa-spinner fa-spin"></i> WorkCom is analyzing automation data...</div>');

            frappe.call({
                method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_ai_insights",
                args: {
                    filters: JSON.stringify(report.get_filter_values()),
                    query: values.query
                },
                callback: function (r) {
                    let answer = (r && r.message && r.message.insights) || "No response received.";
                    $response.html(`<div style="white-space:pre-wrap;">${answer}</div>`);
                },
                error: function () {
                    $response.html('<div class="text-danger">Error getting AI insights. Please try again.</div>');
                }
            });
        }
    });
    d.show();
}

function show_automation_status_chart(report) {
    frappe.call({
        method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_automation_status_chart",
        args: { filters: JSON.stringify(report.get_filter_values()) },
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("Automation Status"), r.message);
            }
        }
    });
}

function show_after_hours_chart(report) {
    frappe.call({
        method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_after_hours_chart",
        args: { filters: JSON.stringify(report.get_filter_values()) },
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("After-Hours Tickets by Platform"), r.message);
            }
        }
    });
}

function show_document_validation_chart(report) {
    frappe.call({
        method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_document_validation_chart",
        args: { filters: JSON.stringify(report.get_filter_values()) },
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("Document Validation Status"), r.message);
            }
        }
    });
}

function show_data_quality_chart(report) {
    frappe.call({
        method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_data_quality_chart",
        args: { filters: JSON.stringify(report.get_filter_values()) },
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("Data Quality Issues"), r.message);
            }
        }
    });
}

function show_ai_failures_chart(report) {
    frappe.call({
        method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_ai_failures_chart",
        args: { filters: JSON.stringify(report.get_filter_values()) },
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("AI Failures by Reason"), r.message);
            }
        }
    });
}

function show_system_health_chart(report) {
    frappe.call({
        method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_system_health_chart",
        args: { filters: JSON.stringify(report.get_filter_values()) },
        callback: function (r) {
            console.log("System health chart response:", r);
            if (r && r.message) {
                show_chart_dialog(__("System Health"), r.message);
            } else {
                frappe.msgprint(__("No data available for system health chart"));
            }
        },
        error: function (err) {
            console.error("System health chart error:", err);
            frappe.msgprint(__("Error loading system health chart"));
        }
    });
}

function show_trend_chart(report) {
    frappe.call({
        method: "construction_v1.construction_v1.report.ai_automation_analysis.ai_automation_analysis.get_automation_trend_chart",
        args: { filters: JSON.stringify(report.get_filter_values()), months: 6 },
        callback: function (r) {
            if (r && r.message) {
                show_chart_dialog(__("Automation Trend (Last 6 Months)"), r.message);
            }
        }
    });
}

function show_chart_dialog(title, chart_data) {
    let d = new frappe.ui.Dialog({
        title: title,
        size: "large"
    });

    // Create chart container
    let chart_container = $('<div class="chart-container" style="height:400px; width:100%;"></div>');
    d.$body.append(chart_container);

    // Show dialog first
    d.show();

    // Use setTimeout to ensure DOM is ready before rendering chart
    setTimeout(function () {
        // Build chart options based on chart type
        let chart_options = {
            data: chart_data.data,
            type: chart_data.type || "bar",
            height: 350,
            colors: chart_data.colors || ["#5e64ff"]
        };

        // Add type-specific options
        if (chart_data.type === "bar" || chart_data.type === "line" || chart_data.type === "axis-mixed") {
            chart_options.barOptions = chart_data.barOptions || {};
            chart_options.lineOptions = chart_data.lineOptions || {};
        }

        // For pie/percentage charts, ensure proper data format
        if (chart_data.type === "pie" || chart_data.type === "percentage") {
            chart_options.maxSlices = chart_data.maxSlices || 10;
        }

        try {
            console.log("Rendering chart with options:", chart_options);
            new frappe.Chart(chart_container[0], chart_options);
        } catch (e) {
            console.error("Chart rendering error:", e);
            chart_container.html('<div class="text-muted text-center" style="padding: 100px;">Unable to render chart: ' + e.message + '</div>');
        }
    }, 100);
}



